### PR TITLE
systemd-detect-virt: refine hypervisor detection

### PR DESCRIPTION
**Pull request by razvan-cojocaru**
Continue to try to get more details about the actual underlying
hypervisor with successive tests until none are available.
This fixes issue #7165.

### CODE REVIEW

File: src/basic/virt.c [hunk 1]
```diff
@@ -387,6 +412,8 @@ int detect_vm(void) {
          * double-check it */
         if (r == VIRTUALIZATION_XEN && detect_vm_xen_dom0())
                 r = VIRTUALIZATION_NONE;
+        else if(r == VIRTUALIZATION_NONE && other)
```
Comment (poettering) [inline src/basic/virt.c]: nitpick: please follow coding style: space between "if" and opening "("...

**Comment by poettering** [comment]
oh, we usually want to merge PRs in "perfect" condition, i.e. where the commits making up the PR reflect logical steps, but not really historical steps. Hence, normally I'd ask you to squash the second patch into the first, but I figure we can do this at the time of merging as well...

**Comment by razvan-cojocaru** [comment]
Right, sorry, thought it would be trivial to squash on merge. Thanks, and if ever there'll be a next time I'll squash and resubmit.

**Comment by poettering** [comment]
hmm, any chance you can rebase this PR on current git and force push? The CI was broken with git until just now, and this was fixed a few moments ago with another commit. To make sure we don't break it right-away again, it would be great if you could rebase this PR and force push so that the CI runs over this again?

**Comment by razvan-cojocaru** [comment]
Done, and squashed the second commit as well.

**Comment by poettering** [comment]
excellent! thanks!

**Comment by razvan-cojocaru** [comment]
No problem, thanks for your promptness.

**Comment by razvan-cojocaru** [comment]
I see a test failed (apparently with a timeout). Should I rebase again?

**Comment by poettering** [comment]
hmm...


```
KVM: entry failed, hardware error 0x0
EAX=00000000 EBX=00000000 ECX=00000000 EDX=000206c1
ESI=00000000 EDI=00000000 EBP=00000000 ESP=00000000
EIP=0000fff0 EFL=00000002 [-------] CPL=0 II=0 A20=1 SMM=0 HLT=0
ES =0000 00000000 0000ffff 00009300
CS =f000 ffff0000 0000ffff 00009b00
SS =0000 00000000 0000ffff 00009300
DS =0000 00000000 0000ffff 00009300
FS =0000 00000000 0000ffff 00009300
GS =0000 00000000 0000ffff 00009300
LDT=0000 00000000 0000ffff 00008200
TR =0000 00000000 0000ffff 00008b00
GDT=     00000000 0000ffff
IDT=     00000000 0000ffff
CR0=60000010 CR2=00000000 CR3=00000000 CR4=00000000
DR0=0000000000000000 DR1=0000000000000000 DR2=0000000000000000 DR3=0000000000000000 
DR6=00000000ffff0ff0 DR7=0000000000000400
EFER=0000000000000000
Code=ff ff 66 5b 66 83 c4 08 66 5b 66 5e 66 c3 b0 20 e6 20 66 c3 <ea> 5b e0 00 f0 30 36 2f 32 33 2f 39 39 00 fc 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
```

that doesn't look like a bug in your code. I figure we should restart the CI. @martinpitt, @keszybz could you do that?

**Comment by razvan-cojocaru** [comment]
I see the test failed again. Could it be that some check that was detecting "other" before now (wrongly) detects KVM and that's what's causing it to fail?

**Comment by poettering** [comment]
Hmm, again:

```
KVM: entry failed, hardware error 0x0
EAX=00000000 EBX=00000000 ECX=00000000 EDX=000306a9
ESI=00000000 EDI=00000000 EBP=00000000 ESP=00000000
EIP=0000fff0 EFL=00000002 [-------] CPL=0 II=0 A20=1 SMM=0 HLT=0
ES =0000 00000000 0000ffff 00009300
CS =f000 ffff0000 0000ffff 00009b00
SS =0000 00000000 0000ffff 00009300
DS =0000 00000000 0000ffff 00009300
FS =0000 00000000 0000ffff 00009300
GS =0000 00000000 0000ffff 00009300
LDT=0000 00000000 0000ffff 00008200
TR =0000 00000000 0000ffff 00008b00
GDT=     00000000 0000ffff
IDT=     00000000 0000ffff
CR0=60000010 CR2=00000000 CR3=00000000 CR4=00000000
DR0=0000000000000000 DR1=0000000000000000 DR2=0000000000000000 DR3=0000000000000000 
DR6=00000000ffff0ff0 DR7=0000000000000400
EFER=0000000000000000
```

It appears we now end up calling some opcode that makes KVM chokes on. Kinda sad that qemu so easy to crash from inside a VM.

**Pull request merged**